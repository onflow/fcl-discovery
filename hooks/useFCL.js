import { useEffect, useState } from 'react'
import { WalletUtils } from '@onflow/fcl'

export function useFCL() {
  const [hasInitialized, setHasInitialized] = useState(false)
  const [loading, setLoading] = useState(false)
  const [appConfig, setAppConfig] = useState()
  const [clientConfig, setClientConfig] = useState()
  const [appVersion, setAppVersion] = useState(null)
  const [extensions, setExtensions] = useState([])
  const [walletInclude, setWalletInclude] = useState([])
  const [wcProjectId, setWcProjectId] = useState(null)
  const [wcPairings, setWcPairings] = useState([])

  useEffect(() => {
    setHasInitialized(true)
    setLoading(true)

    WalletUtils.ready(({ fclVersion, body, config }) => {
      // config.client.fclVersion is only available starting in version 0.0.79-alpha.4
      // config?.client?.extensions starts in fcl v1
      const appFclVersion = config?.client?.fclVersion || fclVersion || null
      const clientExtensions =
        config?.client?.extensions || body?.extensions || []
      const wc = config?.client?.wc || null

      if (config?.app) {
        setAppConfig(config.app)
      }

      if (config?.client) {
        setClientConfig(config.client)
      }

      if (appFclVersion) {
        setExtensions(clientExtensions)
        setAppVersion(appFclVersion)
        setWalletInclude(config.discoveryAuthnInclude || [])
      }

      if (wc?.projectId) {
        setWcProjectId(wc.projectId)
        setWcPairings(wc.pairings)
      }

      setLoading(false)
    })
  }, [])

  return {
    hasInitialized,
    loading,
    appConfig,
    clientConfig,
    appVersion,
    extensions,
    walletInclude,
    wcProjectId,
    wcPairings,
  }
}
