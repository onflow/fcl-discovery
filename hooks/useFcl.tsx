import { useEffect, useRef, useState } from 'react'
import { WalletUtils } from '@onflow/fcl'
import { RpcClient } from '@onflow/util-rpc'
import { Service, Strategy } from '../types'
import { CUSTOM_RPC } from '../helpers/constants'
import {
  DiscoveryNotification,
  FclRequests,
  FclRpcClient,
} from '../helpers/rpc'

type WalletUtilsProps = {
  fclVersion: string
  body: { [key: string]: any }
  config: { [key: string]: any }
}

export interface FclConfig {
  appConfig: { [key: string]: any }
  clientConfig: { [key: string]: any }
  appVersion: string
  walletInclude: string[]
  clientServices: Service[]
  supportedStrategies: Strategy[]
  rpcEnabled?: boolean
}

export function useFcl() {
  const [config, setConfig] = useState<FclConfig | null>(null)
  const [error, setError] = useState<string | null>(null)
  const timeout = useRef<NodeJS.Timeout | null>(null)

  const [rpc, setRpc] = useState<FclRpcClient | null>(null)
  const rpcEnabled = config?.rpcEnabled

  useEffect(() => {
    try {
      WalletUtils.ready(({ fclVersion, body, config }: WalletUtilsProps) => {
        // config.client.fclVersion is only available starting in version 0.0.79-alpha.4
        // config?.client?.extensions starts in fcl v1
        const state = {
          appConfig: config.app,
          clientConfig: config.client,
          appVersion: config.client?.fclVersion || fclVersion || null,
          walletInclude:
            config.discoveryAuthnInclude ||
            config.client.discoveryAuthnInclude ||
            [],
          clientServices:
            config.client?.clientServices ||
            config.client?.extensions ||
            body.extensions ||
            [],
          supportedStrategies: config.client?.supportedStrategies || [],
          rpcEnabled: config.client?.discoveryRpcEnabled || false,
        } as FclConfig

        setConfig(state)
        clearTimeout(timeout.current)
      })

      timeout.current = setTimeout(() => {
        setError(
          'Error occured, if you are the developer please check the console for more information.',
        )
        console.error(
          'Timeout: No response from FCL received within 5s - please ensure the application you are trying to connect to is running FCL & has the correct configuration and try again.',
        )
      }, 5000)
    } catch (e) {
      clearTimeout(timeout.current)
      setError(
        'Error occured, if you are the developer please check the console for more information.',
      )

      console.error(e)
      console.error(
        "An error has occured connecting to the dApp's FCL instance, please see docs: https://developers.flow.com/tools/fcl-js/reference/discovery",
      )
    }

    return () => {
      clearTimeout(timeout.current)
    }
  }, [])

  useEffect(() => {
    if (!rpcEnabled || rpc) return

    const _rpc = new RpcClient<FclRequests, {}>({
      notifications: Object.values(DiscoveryNotification),
    })

    // Bind receiver for messages from FCL
    WalletUtils.onMessageFromFCL(
      CUSTOM_RPC,
      ({ payload: msg }: { payload: any }) => {
        _rpc.receive(msg)
      },
    )

    // Bind sender for messages to FCL
    _rpc.connect({
      send: msg => {
        WalletUtils.sendMsgToFCL(CUSTOM_RPC, { payload: msg })
      },
    })

    setRpc(_rpc)
  }, [rpcEnabled, rpc])

  return {
    error,
    config,
    rpc,
    isLoading: !config && !error,
  }
}
