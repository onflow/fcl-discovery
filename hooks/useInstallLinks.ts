import { useCallback, useMemo } from 'react'
import { Wallet } from '../data/wallets'
import { useDevice } from '../contexts/DeviceContext'
import { DeviceType } from '../helpers/device-info'
import { useConfig } from '../contexts/FclContext'
import { FCL_SERVICE_METHODS } from '../helpers/constants'

export function useInstallLinks(wallet: Wallet) {
  const { deviceInfo } = useDevice()
  const { supportedStrategies } = useConfig()

  const getInstallLink = useCallback(
    (strategy: FCL_SERVICE_METHODS) => {
      let installLink: string | null = null
      switch (strategy) {
        case FCL_SERVICE_METHODS.WC:
          if (deviceInfo.type === DeviceType.MOBILE) {
            installLink =
              wallet.installLink?.[deviceInfo.platform] ||
              wallet.installLink?.mobile
          } else {
            installLink = wallet.installLink?.mobile
          }
          break
        case FCL_SERVICE_METHODS.EXT:
          if (deviceInfo.type !== DeviceType.DESKTOP) {
            return null
          }

          installLink =
            wallet.installLink?.[deviceInfo.browser] ||
            wallet.installLink?.browserExtension

          if (!installLink) {
            return null
          }
          break
      }

      return installLink
    },
    [deviceInfo, wallet],
  )

  const links = useMemo(() => {
    if (!wallet) {
      return {}
    }

    const res = supportedStrategies.reduce(
      (acc, strategy) => {
        let installLink: string | null = getInstallLink(strategy)

        if (!installLink) {
          return acc
        }
        return { ...acc, [strategy]: installLink }
      },
      {} as Record<FCL_SERVICE_METHODS, string>,
    )
    return res
  }, [deviceInfo, wallet, supportedStrategies])

  return links
}
