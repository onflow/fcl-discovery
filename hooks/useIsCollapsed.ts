import { useBreakpointValue } from '@chakra-ui/react'
import { useDevice } from '../contexts/DeviceContext'
import { DeviceType } from '../helpers/device'

export function useIsCollapsed() {
  const { deviceInfo } = useDevice()
  const isCollapsed = useBreakpointValue(
    { base: true, md: false },
    {
      ssr: true,
      fallback:
        deviceInfo.type === DeviceType.MOBILE && !deviceInfo.isTablet
          ? 'md'
          : 'base',
    },
  )
  return isCollapsed
}
