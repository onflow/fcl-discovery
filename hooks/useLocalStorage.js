import { useState, useEffect } from 'react'

function getStorageValue(key, defaultValue) {
  const saved = localStorage.getItem(key)
  const initial = JSON.parse(saved)
  return initial || defaultValue
}

export default function useLocalStorage(key, defaultValue) {
  const [value, setValue] = useState(() => getStorageValue(key, defaultValue))

  useEffect(() => {
    localStorage.setItem(key, JSON.stringify(value))
  }, [key, value])

  return [value, setValue]
}
