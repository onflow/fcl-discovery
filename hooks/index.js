import useFCL from './useFCL'
import useLocalStorage from './useLocalStorage'

export { useFCL, useLocalStorage }
