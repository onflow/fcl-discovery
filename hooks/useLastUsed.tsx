import { LOCAL_STORAGE_KEYS } from '../helpers/constants'
import { useLocalStorage } from './useLocalStorage'

export function useLastUsed() {
  const [, setLastUsed] = useLocalStorage(LOCAL_STORAGE_KEYS.LAST_USED, null)
  const [lastUsed] = useLocalStorage(LOCAL_STORAGE_KEYS.LAST_USED, null)

  return { lastUsed, setLastUsed }
}
