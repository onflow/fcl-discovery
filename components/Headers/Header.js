import FlowHeader from './FlowHeader'
import AppHeader from './AppHeader'
import DeveloperMessage from '../DeveloperMessage'
import { useFCL } from '../../hooks/useFCL'
import { isGreaterThanOrEqualToVersion } from '../../helpers/version'
import { SUPPORTED_VERSIONS } from '../../helpers/constants'
import { isTestnet as isTestnetFn } from '../../helpers/networks'

export default function Header() {
  const isTestnet = isTestnetFn()
  const { appConfig, appVersion } = useFCL()
  const isMissingConfig = !(appConfig?.icon && appConfig?.title)
  const showDeveloperMessage =
    isTestnet &&
    isMissingConfig &&
    isGreaterThanOrEqualToVersion(appVersion, SUPPORTED_VERSIONS.APP_CONFIG)
  const isAppHeaderSupported = isGreaterThanOrEqualToVersion(
    appVersion,
    SUPPORTED_VERSIONS.APP_CONFIG
  )

  return (
    <>
      {isAppHeaderSupported ? (
        <>
          <AppHeader />
          {showDeveloperMessage && <DeveloperMessage />}
        </>
      ) : (
        <FlowHeader />
      )}
    </>
  )
}
