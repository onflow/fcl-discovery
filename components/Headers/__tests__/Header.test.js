import { render } from '@testing-library/react'
import Header from '../Header'
import { useFCL } from '../../../hooks/useFCL'
jest.mock('../../../hooks/useFCL')

jest.mock(
  '../../helpers/networks',
  () => ({
    isTestnet: jest.fn(() => true),
  }),
  { virtual: true }
)

describe('Component: Header', () => {
  test('should render the configurable component if version is old enough', () => {
    useFCL.mockImplementation(() => {
      return {
        appVersion: '1.0.0',
        appConfig: {
          title: 'Test App',
          icon: 'test.png',
        },
        clientConfig: {
          hostname: 'www.onflow.org',
        },
      }
    })

    const { container } = render(<Header />)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('should NOT render the configurable component if version is too low', () => {
    useFCL.mockImplementation(() => {
      return {
        appVersion: '0.0.77',
        appConfig: {
          title: 'Test App',
          icon: 'test.png',
        },
        clientConfig: {
          hostname: 'www.onflow.org',
        },
      }
    })

    const { container } = render(<Header />)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('should show the DeveloperMessage if config is missing', () => {
    useFCL.mockImplementation(() => {
      return {
        appVersion: '1.0.0',
        appConfig: {
          title: null,
          icon: null,
        },
        clientConfig: {
          hostname: null,
        },
      }
    })

    const { container } = render(<Header />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
