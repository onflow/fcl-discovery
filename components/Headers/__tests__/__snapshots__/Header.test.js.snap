// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Header should NOT render the configurable component if version is too low 1`] = `
<div
  class="css-gmuwbf"
>
  <div
    class="css-1j99xrl"
  >
    <img
      alt="Logo"
      class="chakra-image css-kl8a02"
      src="/logo.svg"
    />
    <p
      class="chakra-text css-mwent8"
    >
      Choose a Provider
    </p>
  </div>
</div>
`;

exports[`Component: Header should render the configurable component if version is old enough 1`] = `
<div
  class="chakra-stack css-184jys2"
>
  <h2
    class="chakra-heading css-0"
  >
    Connect to Test App
  </h2>
  <div
    class="chakra-stack css-84zodg"
  >
    <img
      alt="Logo"
      class="chakra-image css-dqjekp"
      src="test.png"
    />
    <p
      class="chakra-text css-mwent8"
    >
      www.onflow.org
    </p>
  </div>
</div>
`;

exports[`Component: Header should show the DeveloperMessage if config is missing 1`] = `
<div
  class="chakra-stack css-184jys2"
>
  <h2
    class="chakra-heading css-0"
  >
    Connect
  </h2>
  <div
    class="chakra-stack css-84zodg"
  >
    <p
      class="chakra-text css-mwent8"
    >
      Unknown
    </p>
  </div>
</div>
`;
